### PR TITLE
Introduce remote document store pluggable component

### PR DESCRIPTION
**Pull request by jonpither**
This builds on PRs #554 and #557 (should be merged first), adding a new remote document store.

This PR will facilitate building an S3 backed remote document store, to work in conjunction with either JDBC or Kafka.

Todos:

- [x] Rename RemoteDocumentStore -> DocumentStore
- [x] Account for #571, stop raising tombstones BEFORE we mark a tx as done.
- [x] Remove open fn of kv-store (or raise issue)

### CODE REVIEW

**Comment by jonpither** [comment]
The underlying issue behind https://github.com/juxt/crux/issues/571 needs to be accounted for - raising tombstones BEFORE we mark a tx as done.

**Comment by jonpither** [comment]
@jarohen and I discussed the following, which I'd consider out the scope of this PR for now.

- [ ] KafkaDocStore, should have a dedicated producer
- [ ] The generation of group-id should just use a uuid,
- [ ] tx-test should submit-docs against node

File: crux-test/test/crux/fixtures/doc_store.clj [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  (:import java.io.Closeable))
+
+(defrecord InMemDocumentStore [docs]
+  Closeable
+  (close [_])
+
+  db/DocumentStore
+  (fetch-docs [this ids]
+    (log/debug "Fetching" (cio/pr-edn-str ids))
+    (into {}
```
Comment (jarohen) [inline crux-test/test/crux/fixtures/doc_store.clj:15]: nit: `(select-keys @docs ids)`, I think - will pop a commit on

File: crux-test/test/crux/fixtures/doc_store.clj [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+          (for [id ids]
+            [id (get @docs id)])))
+
+  (submit-docs [this id-and-docs]
+    (doseq [[content-hash doc] id-and-docs]
+      (log/debug "Storing" (cio/pr-edn-str content-hash))
+      (swap! docs assoc content-hash doc))))
```
Comment (jarohen) [inline crux-test/test/crux/fixtures/doc_store.clj:22]: nit: `(swap! docs merge id-and-docs)`, I think, will add a commit

File: crux-test/test/crux/fixtures/doc_store.clj [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+(def document-store
+  {:start-fn (fn [_ _] (->InMemDocumentStore (atom {})))})
+
+(defn with-remote-doc-store-opts [f]
+  (apif/with-opts {:crux.node/document-store 'crux.fixtures.doc-store/document-store
+                   :crux.kafka/doc-indexing-consumer 'crux.kafka/doc-indexing-from-tx-topic-consumer} f))
```
Comment (jarohen) [inline crux-test/test/crux/fixtures/doc_store.clj]: indentation nit: `f` is the next arg to `with-opts`, shouldn't hide behind the map

File: crux-test/test/crux/fixtures/doc_store.clj (hunk 2 above)
Comment (jarohen) [reply crux-test/test/crux/fixtures/doc_store.clj:22]: ah, these aren't maps. happy to merge, but was a bit of a gotcha as `index-docs` takes a map, IIRC?

**Review by jarohen — APPROVED** [review]
all good, will merge :smile:

**Pull request merged**